File: python/mirage/commands/builtin/generic/tar.py
import io
import tarfile
from collections.abc import Awaitable, Callable

from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _compression_suffix(z: bool, j: bool, J: bool) -> str:
    if z:
        return ":gz"
    if j:
        return ":bz2"
    if J:
        return ":xz"
    return ""


async def _create_archive(
    paths: list[PathSpec],
    archive_path: str,
    mode_suffix: str,
    exclude: str | None,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    accessor: object,
) -> tuple[ByteSource | None, IOResult]:
    filtered = paths
    if exclude:
        filtered = [
            p for p in paths if exclude not in p.original.split("/")[-1]
        ]
    buf = io.BytesIO()
    with tarfile.open(fileobj=buf, mode=f"w{mode_suffix}") as tf:
        for p in filtered:
            data = await read_bytes(accessor, p)
            name = p.original.lstrip("/")
            info = tarfile.TarInfo(name=name)
            info.size = len(data)
            tf.addfile(info, io.BytesIO(data))
    archive = buf.getvalue()
    await write_bytes(accessor, archive_path, archive)
    return None, IOResult(writes={archive_path: archive})


async def _list_archive(
    archive_path: str,
    mode_suffix: str,
    read_bytes: Callable[..., Awaitable[bytes]],
    accessor: object,
) -> tuple[ByteSource | None, IOResult]:
    data = await read_bytes(accessor, archive_path)
    with tarfile.open(fileobj=io.BytesIO(data), mode=f"r{mode_suffix}") as tf:
        names = tf.getnames()
    return ("\n".join(names) + "\n").encode(), IOResult()


async def _extract_archive(
    archive_path: str,
    dest_path: str,
    mode_suffix: str,
    strip_n: int,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    mkdir_fn: Callable[..., Awaitable[None]],
    accessor: object,
) -> tuple[ByteSource | None, IOResult]:
    data = await read_bytes(accessor, archive_path)
    writes: dict[str, bytes] = {}
    with tarfile.open(fileobj=io.BytesIO(data), mode=f"r{mode_suffix}") as tf:
        for member in tf.getmembers():
            if not member.isfile():
                continue
            extracted = tf.extractfile(member)
            if not extracted:
                continue
            content = extracted.read()
            name_parts = member.name.split("/")
            if strip_n > 0:
                name_parts = name_parts[strip_n:]
            if not name_parts:
                continue
            out_path = dest_path.rstrip("/") + "/" + "/".join(name_parts)
            parent = out_path.rsplit("/", 1)[0] or "/"
            if parent != "/":
                await mkdir_fn(accessor, parent, parents=True)
            await write_bytes(accessor, out_path, content)
            writes[out_path] = content
    return None, IOResult(writes=writes)


async def tar(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    mkdir_fn: Callable[..., Awaitable[None]],
    accessor: object = None,
    c: bool = False,
    x: bool = False,
    t: bool = False,
    z: bool = False,
    j: bool = False,
    J: bool = False,
    v: bool = False,
    f: PathSpec | None = None,
    C: PathSpec | None = None,
    strip_components: str | None = None,
    exclude: str | None = None,
) -> tuple[ByteSource | None, IOResult]:
    archive_path = f.strip_prefix if f else None
    dest_path = C.strip_prefix if C else "/"
    mode_suffix = _compression_suffix(z, j, J)
    strip_n = int(strip_components) if strip_components else 0
    if c:
        if not archive_path:
            raise ValueError("tar: -f is required")
        return await _create_archive(paths, archive_path, mode_suffix, exclude,
                                     read_bytes, write_bytes, accessor)
    if t:
        if not archive_path:
            raise ValueError("tar: -f is required")
        return await _list_archive(archive_path, mode_suffix, read_bytes,
                                   accessor)
    if x:
        if not archive_path:
            raise ValueError("tar: -f is required")
        return await _extract_archive(archive_path, dest_path, mode_suffix,
                                      strip_n, read_bytes, write_bytes,
                                      mkdir_fn, accessor)
    raise ValueError("tar: must specify -c, -x, or -t")


__all__ = ["tar"]

File: typescript/packages/core/src/commands/builtin/s3/tar.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { S3Accessor } from '../../../accessor/s3.ts'
import { resolveGlob } from '../../../core/s3/glob.ts'
import { mkdir as s3Mkdir } from '../../../core/s3/mkdir.ts'
import { read as s3Read } from '../../../core/s3/read.ts'
import { write as s3Write } from '../../../core/s3/write.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import { PathSpec, ResourceName } from '../../../types.ts'
import { gzip, gunzip } from '../../../utils/compress.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { readTar, writeTar, type TarEntry } from '../tar_helper.ts'
import { lstripSlash, rstripSlash, stripSlash } from '../../../util/slash.ts'

const ENC = new TextEncoder()

function makePathSpec(original: string, prefix: string): PathSpec {
  return new PathSpec({ original, directory: original, resolved: true, prefix })
}

function fnmatch(name: string, pattern: string): boolean {
  let re = '^'
  for (const ch of pattern) {
    if (ch === '*') re += '.*'
    else if (ch === '?') re += '.'
    else if (/[.+^${}()|[\]\\]/.test(ch)) re += '\\' + ch
    else re += ch
  }
  re += '$'
  return new RegExp(re).test(name)
}

async function compress(data: Uint8Array, z: boolean): Promise<Uint8Array> {
  if (!z) return data
  return gzip(data)
}

async function decompress(data: Uint8Array, z: boolean): Promise<Uint8Array> {
  if (!z) return data
  return gunzip(data)
}

async function tarCommand(
  accessor: S3Accessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const create = opts.flags.c === true
  const extract = opts.flags.x === true
  const list = opts.flags.t === true
  const z = opts.flags.z === true
  const verbose = opts.flags.v === true
  if (opts.flags.j === true || opts.flags.J === true) {
    return [
      null,
      new IOResult({ exitCode: 1, stderr: ENC.encode('tar: bzip2/xz not supported\n') }),
    ]
  }
  const fFlag = typeof opts.flags.f === 'string' ? opts.flags.f : null
  const CFlag = typeof opts.flags.C === 'string' ? opts.flags.C : null
  const stripN =
    typeof opts.flags.strip_components === 'string'
      ? Number.parseInt(opts.flags.strip_components, 10)
      : 0
  const exclude = typeof opts.flags.exclude === 'string' ? opts.flags.exclude : null
  const archivePath = fFlag
  const destPath = CFlag ?? '/'

  const resolved =
    paths.length > 0 ? await resolveGlob(accessor, paths, opts.index ?? undefined) : []
  const mountPrefix = resolved[0]?.prefix ?? ''

  if (create) {
    if (archivePath === null) {
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('tar: -f is required\n') })]
    }
    const filtered =
      exclude !== null
        ? resolved.filter((p) => !fnmatch(p.original.split('/').pop() ?? '', exclude))
        : resolved
    const entries: TarEntry[] = []
    const verboseLines: string[] = []
    for (const p of filtered) {
      const data = await s3Read(accessor, p, opts.index ?? undefined)
      const name = lstripSlash(p.original)
      entries.push({ name, data, isFile: true })
      if (verbose) verboseLines.push(name)
    }
    const archive = await compress(writeTar(entries), z)
    await s3Write(accessor, makePathSpec(archivePath, mountPrefix), archive)
    const stdout = verbose ? ENC.encode(verboseLines.join('\n') + '\n') : null
    return [stdout, new IOResult({ writes: { [archivePath]: archive } })]
  }

  if (list) {
    if (archivePath === null) {
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('tar: -f is required\n') })]
    }
    const archiveSpec = makePathSpec(archivePath, mountPrefix)
    const raw = await s3Read(accessor, archiveSpec, opts.index ?? undefined)
    const data = await decompress(raw, z)
    const entries = readTar(data)
    const out: ByteSource = ENC.encode(entries.map((e) => e.name).join('\n') + '\n')
    return [out, new IOResult()]
  }

  if (extract) {
    if (archivePath === null) {
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('tar: -f is required\n') })]
    }
    const archiveSpec = makePathSpec(archivePath, mountPrefix)
    const raw = await s3Read(accessor, archiveSpec, opts.index ?? undefined)
    const data = await decompress(raw, z)
    const writes: Record<string, Uint8Array> = {}
    const verboseLines: string[] = []
    for (const entry of readTar(data)) {
      if (!entry.isFile) continue
      const nameParts = entry.name.split('/')
      const stripped = stripN > 0 ? nameParts.slice(stripN) : nameParts
      if (stripped.length === 0) continue
      const outPath = rstripSlash(destPath) + '/' + stripped.join('/')
      const parts = stripSlash(outPath).split('/')
      for (let pi = 1; pi < parts.length; pi++) {
        const d = '/' + parts.slice(0, pi).join('/')
        try {
          await s3Mkdir(accessor, makePathSpec(d, mountPrefix))
        } catch {
          // already exists
        }
      }
      await s3Write(accessor, makePathSpec(outPath, mountPrefix), entry.data)
      writes[outPath] = entry.data
      if (verbose) verboseLines.push(entry.name)
    }
    const stdout = verbose ? ENC.encode(verboseLines.join('\n') + '\n') : null
    return [stdout, new IOResult({ writes })]
  }

  return [
    null,
    new IOResult({ exitCode: 1, stderr: ENC.encode('tar: must specify -c, -x, or -t\n') }),
  ]
}

export const S3_TAR = command({
  name: 'tar',
  resource: ResourceName.S3,
  spec: specOf('tar'),
  fn: tarCommand,
  write: true,
})
